tests/src/test/java/alluxio/server/ft/journal/raft/EmbeddedJournalIntegrationTestFaultTolerance.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.server.ft.journal.raft;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertNotNull;
+
+import alluxio.AlluxioURI;
+import alluxio.Constants;
+import alluxio.client.file.FileSystem;
+import alluxio.conf.PropertyKey;
+import alluxio.exception.FileAlreadyExistsException;
+import alluxio.exception.FileDoesNotExistException;
+import alluxio.master.journal.JournalType;
+import alluxio.master.journal.raft.RaftJournalSystem;
+import alluxio.master.journal.raft.RaftJournalUtils;
+import alluxio.multi.process.MultiProcessCluster;
+import alluxio.multi.process.PortCoordination;
+import alluxio.util.CommonUtils;
+import alluxio.util.WaitForOptions;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.ratis.server.RaftServerConfigKeys;
+import org.apache.ratis.server.storage.RaftStorageImpl;
+import org.apache.ratis.statemachine.impl.SimpleStateMachineStorage;
+import org.apache.ratis.statemachine.impl.SingleFileSnapshotInfo;
+import org.junit.Test;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class EmbeddedJournalIntegrationTestFaultTolerance
+    extends EmbeddedJournalIntegrationTestBase {
+
+  private static final int RESTART_TIMEOUT_MS = 6 * Constants.MINUTE_MS;
+  private static final int NUM_MASTERS = 3;
+  private static final int NUM_WORKERS = 0;
+
+  @Test
+  public void failover() throws Exception {
+    mCluster = MultiProcessCluster.newBuilder(PortCoordination.EMBEDDED_JOURNAL_FAILOVER)
+        .setClusterName("EmbeddedJournalFaultTolerance_failover")
```

Comment:
potential to automatically generate the name using the class name and then getting the calling method name: https://www.geeksforgeeks.org/get-name-of-current-method-being-executed-in-java/

Reply:
I wanted to refactor these cluster setup lines into one `@Before` method, but that would cause problems with `PortCoordination` as we've discussed before, so I made all the name explicit.

Reply:
I understand that. I just meant you could do something like:

```java
...
      .setClusterName(generateClusterName())
...

public String generateClusterName() {
  return this.class.getName() + "_" + new Throwable().getStackTrace()[1].getMethodName();
}
```

Reply:
There's actually a simpler way of doing this in JUnit:
```
@Rule 
public TestName testName = new TestName();
...
 testName.getMethodName();
```
This gets you the name of the test that is currently running, even in `@Before` methods.

Reply:
:O never knew this. Nice!